concurrency_limiter_test.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+type fakeHandler struct {
+	ctx                         context.Context
+	completeInFlightRequestChan chan struct{}
+}
+
+func (f *fakeHandler) ServerHTTP(w http.ResponseWriter, r *http.Request) {
+
+	select {
+	case <-f.ctx.Done():
+		w.WriteHeader(http.StatusServiceUnavailable)
+	case <-f.completeInFlightRequestChan:
+		w.WriteHeader(http.StatusOK)
+	}
+
+}
+
+func doRRandRequest(ctx context.Context, t *testing.T, wg *sync.WaitGroup, handler http.Handler, fh *fakeHandler) *httptest.ResponseRecorder {
```

Comment:
Is it worth returning the error so that testing.T doesn't have to be a dependency of the method?